x/concentrated-liquidity/position.go

```diff
@@ -664,3 +665,102 @@ func (k Keeper) updateFullRangeLiquidityInPool(ctx sdk.Context, poolId uint64, l
+// If the sender is not the owner, it returns an error.
+// It then checks if the position has an active underlying lock, and if so, returns an error.
+// It then collects any outstanding incentives and rewards for the position, deletes the KVStore entries for the position,
+// and restores the position under the recipient's account.
+// If any of these operations fail, it returns the corresponding error.
+// If all operations succeed, it returns nil.
+func (k Keeper) transferPositions(ctx sdk.Context, positionIds []uint64, sender sdk.AccAddress, recipient sdk.AccAddress) error {
+	// All position IDs in the array must be unique.
+	if !osmoassert.Uint64ArrayValuesAreUnique(positionIds) {
+		return types.DuplicatePositionIdsError{PositionIds: positionIds}
+	}
+
+	for _, positionId := range positionIds {
+		position, err := k.GetPosition(ctx, positionId)
+		if err != nil {
+			return err
+		}
+
+		// Check that the sender is the owner of the position.
+		if position.Address != sender.String() {
+			return types.PositionOwnerMismatchError{PositionOwner: position.Address, Sender: sender.String()}
+		}
+
+		// If the position has an active underlying lock, we cannot transfer it.
+		positionHasActiveUnderlyingLock, lockId, err := k.positionHasActiveUnderlyingLockAndUpdate(ctx, positionId)
+		if err != nil {
+			return err
+		}
+		if positionHasActiveUnderlyingLock {
+			return types.LockNotMatureError{PositionId: position.PositionId, LockId: lockId}
+		}
+
+		// Collect any outstanding incentives and rewards for the position.
+		if _, err := k.collectSpreadRewards(ctx, sender, positionId); err != nil {
+			return err
+		}
+		if _, _, err := k.collectIncentives(ctx, sender, positionId); err != nil {
+			return err
+		}
+
+		// Delete the KVStore entries for the position.
+		err = k.deletePosition(ctx, positionId, sender, position.PoolId)
+		if err != nil {
+			return err
+		}
+
+		// There exists special logic branches we take if a position is the last position in a pool and it is withdrawn.
+		// It makes sense to accept the small annoyance of preventing a position from being transferred if it is the last position in a pool
+		// instead of considering all the edge cases that would arise if we allowed it.
+		anyPositionsRemainingInPool, err := k.HasAnyPositionForPool(ctx, position.PoolId)
+		if err != nil {
+			return err
+		}
+		if !anyPositionsRemainingInPool {
+			return types.LastPositionTransferError{PositionId: positionId, PoolId: position.PoolId}
+		}
+
+		// Restore the position under the recipient's account.
+		err = k.SetPosition(ctx, position.PoolId, recipient, position.LowerTick, position.UpperTick, position.JoinTime, position.Liquidity, position.PositionId, 0)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// underlyingPositionsValue calculates the value of the underlying assets in the given positions.
+func (k Keeper) UnderlyingPositionsValue(ctx sdk.Context, positionIds []uint64) (sdk.Coins, error) {
```

Comment:
Since this function is only used in tests, I'd be in favor of moving it to the test file unless we expect to use this in production code in the future

Reply:
We expect this the be used in a query as mentioned here https://github.com/osmosis-labs/osmosis/pull/6623#discussion_r1344814214

I will make an issue

Reply:
Issue made https://github.com/osmosis-labs/osmosis/issues/6650